File: app/src/main/java/com/example/loren/minesample/ChatActivity.java
package com.example.loren.minesample;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.transition.Explode;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.example.loren.minesample.widget.AdjustLinearLayout;
import com.facebook.drawee.backends.pipeline.Fresco;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

import bean.MessageBean;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import util.Constant;

public class ChatActivity extends AppCompatActivity implements AdjustLinearLayout.onSizeChangeListener {

    @BindView(R.id.chat_rv)
    RecyclerView chatRv;
    @BindView(R.id.input_edt)
    EditText inputEdt;
    @BindView(R.id.send_btn)
    Button sendBtn;
    @BindView(R.id.title_tv)
    TextView titleTv;
    @BindView(R.id.adjust_ll)
    AdjustLinearLayout adjustLl;
    private Disposable disposable;
    private ArrayList<MessageBean.DataBean> data = null;
    private ChatAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fresco.initialize(this);
        setContentView(R.layout.activity_chat);
        ButterKnife.bind(this);
        setAnim();
        Observable.interval(10, TimeUnit.SECONDS, Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Long>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        disposable = d;
                    }

                    @Override
                    public void onNext(Long aLong) {
                        MessageBean.DataBean bean = new MessageBean.DataBean();
                        bean.setId("");
                        bean.setName("");
                        bean.setMsg("嗯！");
                        bean.setType("1");
                        bean.setUrl("http://tupian.enterdesk.com/2014/lxy/2014/12/01/5/1.jpg");
                        //refresh(bean);
                        Observable.just(bean)
                                .subscribe(new Consumer<MessageBean.DataBean>() {
                                    @Override
                                    public void accept(MessageBean.DataBean bean) throws Exception {
                                        refresh(bean);
                                    }
                                });
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });

        data = new Gson().fromJson(Constant.JSON, MessageBean.class).getData();
        titleTv.setText(data.get(1).getName());
        mAdapter = new ChatAdapter(this, data);

        chatRv.setLayoutManager(new GridLayoutManager(this, 1));
        chatRv.setAdapter(mAdapter);
        chatRv.getAdapter().notifyDataSetChanged();

        sendBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (inputEdt.getText().toString().trim().equals(""))
                    return;
                MessageBean.DataBean bean = new MessageBean.DataBean();
                bean.setId("");
                bean.setName("");
                bean.setMsg(inputEdt.getText().toString());
                bean.setType("0");
                bean.setUrl("http://img1.imgtn.bdimg.com/it/u=648722154,281408629&fm=11&gp=0.jpg");
                refresh(bean);
                inputEdt.setText("");
            }
        });
        adjustLl.setListener(this);
    }

    private void setAnim() {
        getWindow().setEnterTransition(new Explode().setDuration(1000));
        getWindow().setExitTransition(new Explode().setDuration(1000));
    }

    private void refresh(MessageBean.DataBean bean) {
        int temp = data.size();
        data.add(bean);
        chatRv.getAdapter().notifyItemRangeInserted(temp, data.size());
        chatRv.scrollBy(0, Integer.MAX_VALUE);
    }

    @Override
    protected void onDestroy() {
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
        super.onDestroy();
    }

    @Override
    public void onSizeChange(boolean isShowInputMethod) {
        if (isShowInputMethod)
            chatRv.post(new Runnable() {
                @Override
                public void run() {
                    chatRv.scrollBy(0, Integer.MAX_VALUE);
                }
            });

    }
}

File: app/src/main/java/com/example/loren/minesample/ChatActivity.kt
package com.example.loren.minesample

import android.support.v7.widget.GridLayoutManager
import android.transition.Explode
import android.view.View
import bean.MessageBean
import com.example.loren.minesample.base.ui.BaseActivity
import com.example.loren.minesample.util.Constant
import com.example.loren.minesample.util.http
import com.example.loren.minesample.widget.AdjustLinearLayout
import com.google.gson.Gson
import kotlinx.android.synthetic.main.activity_chat.*
import pers.victor.ext.toast
import java.util.*

class ChatActivity : BaseActivity(), AdjustLinearLayout.onSizeChangeListener {
    //ALI interface 2018.10.30
    private val ANSWER_URL = "http://jisuznwd.market.alicloudapi.com/iqa/query?question=%s"
    //SecretId: AKIDZzCBxVixrX1ILAs9JGxOcaIpAF4eM1oK
    //SecretKey: QbXQ7YjeNq30qSpxrWx85nGUTDm6B2e1
    private val TECENT_URL = "https://wenzhi.api.qcloud.com/v2/index.php?" +
            "    Action=TextSentiment" +
            "    &Nonce=345122" +
            "    &Region=qingdao" +
            "    &SecretId=AKIDZzCBxVixrX1ILAs9JGxOcaIpAF4eM1oK" +
            "    &Timestamp=1408704141" +
            "    &Signature=HgIYOPcx5lN6gz8JsCFBNAWp2oQ" +
            "    &content=哈哈哈"

    private fun getAnswer(question: String) {
        http {
            url = String.format(ANSWER_URL, question)
            success {
                addAnswerBean(it.toString())
            }
            fail { toast(it.toString()) }
        }
    }

    override fun initWidgets() {
        setAnim()
        data = Gson().fromJson(Constant.JSON, MessageBean::class.java).data
        title_tv!!.text = "智能问答"
        mAdapter = ChatAdapter(this, data!!)

        chat_rv!!.layoutManager = GridLayoutManager(this, 1)
        chat_rv!!.adapter = mAdapter
        chat_rv!!.adapter.notifyDataSetChanged()

        send_btn!!.setOnClickListener(View.OnClickListener {
            if (input_edt!!.text.toString().trim { it <= ' ' } == "")
                return@OnClickListener
            val bean = MessageBean.DataBean()
            bean.id = ""
            bean.name = ""
            bean.msg = input_edt!!.text.toString()
            bean.type = "0"
            bean.url = "http://img1.imgtn.bdimg.com/it/u=648722154,281408629&fm=11&gp=0.jpg"
            refresh(bean)
            getAnswer(input_edt!!.text.toString())
            input_edt!!.setText("")
        })
    }

    private fun addAnswerBean(msg: String) {
        val bean = MessageBean.DataBean()
        bean.id = ""
        bean.name = ""
        bean.msg = msg
        bean.type = "1"
        bean.url = "http://awb.img.xmtbang.com/img/uploadnew/201602/25/02512de9a2874184873c56dadc6dbcbd.jpg"
        refresh(bean)
    }

    override fun setListeners() {
        click(adjust_ll)
    }

    override fun onWidgetsClick(v: View) {
    }

    override fun bindLayout() = R.layout.activity_chat

    private var data: ArrayList<MessageBean.DataBean>? = null
    private var mAdapter: ChatAdapter? = null


    private fun setAnim() {
        window.enterTransition = Explode().setDuration(200)
        window.exitTransition = Explode().setDuration(200)
    }

    private fun refresh(bean: MessageBean.DataBean) {
        val temp = data!!.size
        data!!.add(bean)
        chat_rv!!.adapter.notifyItemRangeInserted(temp, data!!.size)
        chat_rv!!.scrollBy(0, Integer.MAX_VALUE)
    }

    override fun onSizeChange(isShowInputMethod: Boolean) {
        if (isShowInputMethod)
            chat_rv!!.post { chat_rv!!.scrollBy(0, Integer.MAX_VALUE) }
    }

    override fun useTitleBar() = false
}
